Bluetooth: controller: Implement CSA #2 for SubEvents

cvinayak opened the pull request:
Implement the functions required to calculate the SubEvent 1
and SubEvent n mapped channel indices.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Implement CSA #2 for SubEvents

Implement the functions required to calculate the SubEvent 1
and SubEvent n mapped channel indices.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented:
@wopu-ot please review.

wopu-ot commented:
@heml-ot @mtpr-ot please comment on how this CSA#2 implementation squares with our local implementation.

Thalley commented on subsys/bluetooth/controller/ll_sw/lll_chan.c:
```diff
@@ -158,20 +194,305 @@ static uint16_t chan_mam(uint16_t a, uint16_t b)
+
+	m = lll_chan_sel_2(8, 0x305F, chan_map_2, 9);
+	LL_ASSERT(m == 34U);
+
+
+
```
I guess we could reduce this to a single empty line

cvinayak replied:
done

Thalley commented on subsys/bluetooth/controller/ll_sw/lll_chan.c:
```diff
@@ -158,20 +194,305 @@ static uint16_t chan_mam(uint16_t a, uint16_t b)
+	uint16_t prn_s;
+	uint8_t m;
+
+	m = lll_chan_sel_2(0, 0x305F, chan_map_1, 37);
```
What is the `0x305F` value? It seems to be used everywhere; consider doing a `#define` for it, or at least a const variable.

cvinayak replied:
done.

Thalley commented on subsys/bluetooth/controller/ll_sw/lll_chan.c:
```diff
@@ -158,20 +194,305 @@ static uint16_t chan_mam(uint16_t a, uint16_t b)
+
+#if defined(CONFIG_BT_CTLR_ADV_ISO) || defined(CONFIG_BT_CTLR_SYNC_ISO)
+	/* BIS subevent 2, event counter 0, test prnSubEvent_se */
+	prn_s = 56857 ^ 0x305F;
```
I'm going to assume that the `56857` value, as well as the `0x305F` value, are defined by some specification. Consider defining `56857` in a const variable or as a `#define`. 

It's also somewhat odd that one is defined as a hex value, and the other as an base-10 value.

cvinayak replied:
This is how it is documented in the BT Spec sample data, and will keep it that way.

Thalley commented on subsys/bluetooth/controller/ll_sw/lll_chan.c:
```diff
@@ -158,20 +194,305 @@ static uint16_t chan_mam(uint16_t a, uint16_t b)
+	/* BIS subevent 4, event counter 0 */
+	prn_subevent_se = chan_prn_subevent_se(0x305F, &prn_subevent_lu);
+	LL_ASSERT(prn_subevent_se == 38198);
+
+
+
```
Reduce to a single empty line?

cvinayak replied:
Done

Thalley commented on subsys/bluetooth/controller/ll_sw/lll_chan.c:
```diff
@@ -158,20 +194,305 @@ static uint16_t chan_mam(uint16_t a, uint16_t b)
+#endif /* CONFIG_BT_CTLR_ADV_ISO || CONFIG_BT_CTLR_SYNC_ISO */
+
+#if defined(CONFIG_BT_CTLR_TEST)
+void lll_chan_sel_2_ut(void)
```
It somewhat makes sense to have this as a single function, although I think there might be some readability to be gained by splitting it out. It seems that it's already logically separated using empty lines - I think separation by functions is a better solution.

cvinayak replied:
Tests are at best in the order close to what I could translate the sample data table in the BT Spec, table 3.1 and table 3.2. Refactoring will make it difficult to relate to the documented values and order of inputs used for testing.

kruithofa commented on subsys/bluetooth/controller/ll_sw/lll_chan.c:
```diff
@@ -158,20 +194,305 @@ static uint16_t chan_mam(uint16_t a, uint16_t b)
+}
+#endif /* CONFIG_BT_CTLR_ADV_ISO || CONFIG_BT_CTLR_SYNC_ISO */
+
+#if defined(CONFIG_BT_CTLR_TEST)
```
I propose to make these standard unit tests, so that they can be run by invoking twister

cvinayak replied:
In order to test static function and to keep the test close to the implementation, this was convenient and met the validation. Have added it to babblesim ISO test.

kruithofa commented on subsys/bluetooth/controller/ll_sw/lll_chan.c:
```diff
@@ -158,20 +194,305 @@ static uint16_t chan_mam(uint16_t a, uint16_t b)
+	prn_s = 56857 ^ 0x305F;
+	prn_subevent_lu = prn_s;
+	prn_subevent_se = chan_prn_subevent_se(0x305F, &prn_subevent_lu);
+	LL_ASSERT(prn_subevent_se == 11710);
```
refer to the BT spec where these numbers come from

cvinayak replied:
Done

kruithofa commented on subsys/bluetooth/controller/ll_sw/lll_chan.c:
```diff
@@ -158,20 +194,305 @@ static uint16_t chan_mam(uint16_t a, uint16_t b)
+{
+	uint8_t x, y;
+
+	if (n > 5) {
```
Are these if statements really necessary? Unless I'm misreading the MIN and MAX functions ensure that we get the proper values in any case

cvinayak committed:
tests: Bluetooth: Controller: Cover in-system tests in CI

Include the in-system tests Kconfig option in build-only and
babblesim tests.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Replace test vector inputs with auto variable

Replace hard coded channel map count and channel identifier
with an auto variables.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Added comments for CSA #2 implementation

Added Bluetooth Specification references to the
implementation of Channel Selection algorithm #2 in the
Controller.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/ll_sw/lll_chan.c:
```diff
@@ -17,10 +17,21 @@
+static uint16_t chan_prn_s(uint16_t counter, uint16_t chan_id);
+static uint16_t chan_prn_e(uint16_t counter, uint16_t chan_id);
+
+#if defined(CONFIG_BT_CTLR_ADV_ISO) || defined(CONFIG_BT_CTLR_SYNC_ISO)
```
```suggestion
#if defined(CONFIG_BT_CTLR_BROADCAST_ISO)
```

Thalley commented on subsys/bluetooth/controller/ll_sw/lll_chan.c:
```diff
@@ -73,8 +91,81 @@ uint8_t lll_chan_sel_2(uint16_t counter, uint16_t chan_id, uint8_t *chan_map,
 	return chan_next;
 }
+
+#if defined(CONFIG_BT_CTLR_ADV_ISO) || defined(CONFIG_BT_CTLR_SYNC_ISO)
```
```suggestion
#if defined(CONFIG_BT_CTLR_BROADCAST_ISO)
```

Thalley commented on subsys/bluetooth/controller/ll_sw/lll_chan.c:
```diff
@@ -158,20 +225,329 @@ static uint16_t chan_mam(uint16_t a, uint16_t b)
 	return prn_e;
 }
+
+#if defined(CONFIG_BT_CTLR_ADV_ISO) || defined(CONFIG_BT_CTLR_SYNC_ISO)
```
```suggestion
#if defined(CONFIG_BT_CTLR_BROADCAST_ISO)
```

Thalley reviewed (approved):
LGTM. Minor suggestion to use `CONFIG_BT_CTLR_BROADCAST_ISO`, but not necessary.

cvinayak committed:
Bluetooth: Controller: Use BT_CTLR_BROADCAST_ISO Kconfig option

Use CONFIG_BT_CTLR_BROADCAST_ISO when CONFIG_BT_CTLR_ADV_ISO
or CONFIG_BT_CTLR_SYNC_ISO Kconfig option is supported.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented:
@kruithofa please re-review.
